nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java

```diff
@@ -147,52 +160,149 @@ private TempBlob transformXmlMetadata(final TempBlob artifact,
+        return Collections.singletonList(urlString);
+      }
+      else {
+        List<String> simpleRepositories = new ArrayList<>();
+        TempBlob tempBlob = convertFileToTempBlob(compositeContentTempFile.get(), repository);
+        try (InputStream xmlIn = xmlInputStream(tempBlob, file + ".xml",
+            FilenameUtils.getExtension(compositeContentTempFile.get().toString()), compositeContentTempFile.get())) {
+          XMLInputFactory inputFactory = XMLInputFactory.newFactory();
+          XMLEventReader reader = null;
+          try {
+            reader = inputFactory.createXMLEventReader(xmlIn);
+            while (reader.hasNext()) {
+              XMLEvent event = reader.nextEvent();
+
+              if (isStartTagWithName(event, "child")) {
+                Optional<String> locationAttributeFromXmlEvent = getLocationAttributeFromXmlEvent((StartElement) event);
+                if (locationAttributeFromXmlEvent.isPresent()) {
+                  simpleRepositories.addAll(
+                      changeLocationAttribute(locationAttributeFromXmlEvent.get(), URI.create(urlString), repository, file));
+                }
+              }
+            }
+          }
+          finally {
+            if (reader != null) {
+              reader.close();
+            }
+          }
+        }
+        return simpleRepositories;
+      }
+    }
+    catch (MalformedURLException e) {
+      log.error("Invalid composite repository: {} ", urlString);
     }
+    finally {
+      if (compositeContentTempFile.get() != null) {
+        delete(compositeContentTempFile.get());
+      }
+    }
+    return Collections.emptyList();
+  }
+
+  private Optional<Path> getFilePath(final String baseUrl, final String filename, final String extension) throws IOException {
+    Path tempFile = null;
+    URL url = new URL(baseUrl + filename + extension);
+    boolean isRemoteExist = isRemoteFileExist(url);
+    if (isRemoteExist) {
+      InputStream in = url.openStream();
+      tempFile = createTempFile(filename, extension);
+      Files.copy(in, tempFile, StandardCopyOption.REPLACE_EXISTING);
+    }
+
+    return Optional.ofNullable(tempFile);
+  }
 
-    String value = locationAttribute.getValue();
-    URI uri = URI.create(value);
-    String assetPath =
-        P2PathUtils.escapeUriToPath(uri.isAbsolute() ? uri.toString() : remoteUrl.resolve(uri).toString());
-    String locationValue = String.format(REPOSITORY, nexusRepositoryName, assetPath);
-    StartElement startElement =
-        XMLEventFactory.newInstance().createStartElement(new QName("child"), Collections.singletonList(
-            XMLEventFactory.newInstance()
-                .createAttribute(locationAttrName, locationValue))
-                .iterator(),
-            Collections.emptyIterator());
-    return startElement;
+  private boolean isRemoteFileExist(final URL url) throws IOException {
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```

Comment:
HttpClient client = getRepository().facet(HttpClientFacet.class).getHttpClient();

Reply:
CloseableHttpClient client = HttpClients.createMinimal() will be better